### PR TITLE
feat(settings): allow partial presets (@amarnathsama)

### LINKED ISSUE #4532: Allow partial presets

### PR DESCRIPTION
**Pull request by amarnathsama**
### Description
Allows creation of create partial presets.

Key Points-
1) For active setting groups, all settings including defaults are saved and set.
2) For inactive groups, settings are unchanged when preset is applied.

### Checks

- [x] Check if any open issues are related to this PR; if so, be sure to tag them below.
- [x] Make sure the PR title follows the Conventional Commits standard. (https://www.conventionalcommits.org for more info)
- [x] Make sure to include your GitHub username prefixed with @ inside parentheses at the end of the PR title.

Closes #4532

### CODE REVIEW

File: frontend/src/ts/config.ts [hunk 1]
```diff
@@ -2081,4 +2081,20 @@ export const loadPromise = new Promise((v) => {
   loadDone = v;
 });
 
+export async function selectiveApply(
```
Comment (fehmer) [inline frontend/src/ts/config.ts:2084]: could you add tests for this?

File: packages/contracts/src/schemas/presets.ts [hunk 2]
```diff
@@ -8,11 +8,47 @@ export const PresetNameSchema = z
   .max(16);
 export type PresentName = z.infer<typeof PresetNameSchema>;
 
+export const presetSettingGroupSchema = z.enum([
```
Comment (fehmer) [inline packages/contracts/src/schemas/presets.ts:11]: please start schemas with uppercase letter

File: packages/contracts/src/schemas/presets.ts [hunk 3]
```diff
@@ -8,11 +8,47 @@ export const PresetNameSchema = z
...
+  "test",
+  "account",
+  "behaviour",
+  "input",
+  "sound",
+  "caret",
+  "appearance",
+  "theme",
+  "hide elements",
```
Comment (fehmer) [inline packages/contracts/src/schemas/presets.ts:20]: i'd use `hideElements`

File: packages/contracts/src/schemas/presets.ts [hunk 4]
```diff
@@ -8,11 +8,47 @@ export const PresetNameSchema = z
...
+    );
+  });
+export type ActiveSettingGroups = z.infer<typeof activeSettingGroupsSchema>;
+
 export const PresetSchema = z.object({
   _id: IdSchema,
   name: PresetNameSchema,
   config: PartialConfigSchema.extend({
     tags: z.array(TagSchema).optional(),
+    settingGroups: activeSettingGroupsSchema,
```
Comment (fehmer) [inline packages/contracts/src/schemas/presets.ts:51]: i guess this should be optional. if all groups are selected have it empty to keep it the same as current "full" presets.

File: packages/contracts/src/schemas/presets.ts [hunk 5]
```diff
@@ -8,11 +8,47 @@ export const PresetNameSchema = z
...
+  "appearance",
+  "theme",
+  "hide elements",
+  "ads",
+]);
+export type PresetSettingGroup = z.infer<typeof presetSettingGroupSchema>;
+export const activeSettingGroupsSchema = z
+  .array(presetSettingGroupSchema)
+  .min(1)
+  .superRefine((settingList, ctx) => {
```
Comment (fehmer) [inline packages/contracts/src/schemas/presets.ts:27]: please add a test case in `controllers/preset.spec.ts`

File: packages/contracts/src/schemas/presets.ts (hunk 4 above)
Comment (amarnathsama) [reply packages/contracts/src/schemas/presets.ts:51]: keeping it undefined makes a preset indistinguishable from a legacy preset, which we need to be able to determine to migrate legacy values, how about i keep the list empty for full?

File: frontend/src/ts/modals/edit-preset.ts [hunk 6]
```diff
@@ -158,12 +239,165 @@ async function apply(): Promise<void> {
...
+    case "customTheme":
+    case "customThemeColors":
+    case "favThemes":
+    case "flipTestColors":
+    case "colorfulMode":
+    case "randomTheme":
+    case "customBackground":
+    case "customBackgroundSize":
+    case "customBackgroundFilter":
+      return presetSettingGroupSchema.Enum.theme;
```
Comment (fehmer) [inline frontend/src/ts/modals/edit-preset.ts:257]: you can just return the value as string, e.g. `"theme"` typescript will check this is a valid enum value

File: backend/__tests__/api/controllers/preset.spec.ts
Comment (fehmer) [inline backend/__tests__/api/controllers/preset.spec.ts:1]: i'm not sure about this changes. I guess we will still need to support "legacy"/full presets. You can edit existing presets on the settings page and apply the current settings to them.

In addition we should avoid breaking changes if possible so we don't cause issues in old clients.

File: packages/contracts/src/schemas/presets.ts (hunk 2 above)
Comment (amarnathsama) [reply packages/contracts/src/schemas/presets.ts]: Updated

File: packages/contracts/src/schemas/presets.ts (hunk 3 above)
Comment (amarnathsama) [reply packages/contracts/src/schemas/presets.ts]: Updated

File: packages/contracts/src/schemas/presets.ts (hunk 4 above)
Comment (amarnathsama) [reply packages/contracts/src/schemas/presets.ts]: Updated

File: frontend/src/ts/modals/edit-preset.ts (hunk 6 above)
Comment (amarnathsama) [reply frontend/src/ts/modals/edit-preset.ts]: fixed

File: backend/__tests__/api/controllers/preset.spec.ts
Comment (amarnathsama) [reply backend/__tests__/api/controllers/preset.spec.ts:1]: fixed

File: packages/contracts/src/schemas/presets.ts (hunk 5 above)
Comment (amarnathsama) [reply packages/contracts/src/schemas/presets.ts:30]: done

File: frontend/src/ts/config.ts (hunk 1 above)
Comment (amarnathsama) [reply frontend/src/ts/config.ts:2084]: done

**Review by Miodec — CHANGES_REQUESTED** [review]
One more thing i found - while editing a preset, the full/partial option should only show if you check "change preset to current settings"

File: frontend/src/ts/modals/edit-preset.ts [hunk 7]
```diff
@@ -158,12 +306,186 @@ async function apply(): Promise<void> {
...
+    case "monkeyPowerLevel":
+      return "hidden";
+
+    case "ads":
+      return "ads";
+    default:
+      break;
+  }
+  throw new Error(`${configFieldName} setting not part of any setting group`);
+}
```
[the remark this answers is not in the document]
Comment (amarnathsama) [reply frontend/src/ts/modals/edit-preset.ts]: done

[commit link](https://github.com/monkeytypegame/monkeytype/pull/5813/commits/78e549e7f83ee8736cfd06cc21ed5f1a81567f4f)

please let me know if this is good.

File: frontend/src/ts/modals/edit-preset.ts [hunk 8]
```diff
@@ -19,42 +38,152 @@ export function show(action: string, id?: string, name?: string): void {
...
+      rowElem = rowElem.concat(`
+            <div class="checkboxTitlePair" data-id="${nextSettingGroup}">
+              <input type="checkbox" />
+              <div class="title">${nextSettingGroupTitle}</div>
+          </div>`);
+    }
+    settingGroupListEl.append(
+      `<div class="checkboxGroupRow">` + rowElem + `<div>`
+    );
+  }
```
[the remark this answers is not in the document]
Comment (amarnathsama) [reply frontend/src/ts/modals/edit-preset.ts]: done

**Comment by amarnathsama** [comment]
> One more thing i found - while editing a preset, the full/partial option should only show if you check "change preset to current settings"

done

File: packages/contracts/src/schemas/presets.ts [hunk 9]
```diff
@@ -8,11 +8,56 @@ export const PresetNameSchema = z
...
   _id: IdSchema,
   name: PresetNameSchema,
+  settingGroups: ActiveSettingGroupsSchema.optional(),
   config: PartialConfigSchema.extend({
     tags: z.array(TagSchema).optional(),
   }),
 });
 export type Preset = z.infer<typeof PresetSchema>;
+
+export const EditPresetRequestSchema = PresetSchema.extend({
```
Comment (fehmer) [inline packages/contracts/src/schemas/presets.ts:59]:
```ts
export const EditPresetRequestSchema = PresetSchema.partial({
  config:true
})```

File: backend/src/dal/preset.ts [hunk 10]
```diff
@@ -56,11 +59,14 @@ export async function addPreset(
   };
 }
 
-export async function editPreset(uid: string, preset: Preset): Promise<void> {
+export async function editPreset(
```
Comment (fehmer) [inline backend/src/dal/preset.ts:62]: i'd implement it this way:
```ts
export async function editPreset(uid: string, preset: Preset): Promise<void> {
  const update: Partial<Omit<Preset, "_id">> = omit(preset, "_id");

  if (
    preset.config === undefined ||
    preset.config === null ||
    Object.keys(preset.config).length === 0
  ) {
    delete update.config;
  }

  await getPresetsCollection().updateOne(getPresetKeyFilter(uid, preset._id), {
    $set: update,
  });
}
```

Comment (amarnathsama) [reply backend/src/dal/preset.ts:63]: fixed

**Review by Miodec — DISMISSED** [review]

File: backend/__tests__/dal/preset.spec.ts [hunk 11]
```diff
@@ -219,6 +266,151 @@ describe("PresetDal", () => {
...
+          settingGroups: ["hideElements"],
+          config: {
+            showKeyTips: true,
+            capsLockWarning: true,
+            showOutOfFocusWarning: true,
+            showAverage: "off",
+          },
+        })
+      ).presetId;
+      const second = (
```
[the remark this answers is not in the document]
Comment (fehmer) [reply backend/__tests__/dal/preset.spec.ts:285]: Example:
```ts
 it("should edit when partial is edited to full", async () => {
      //GIVEN
      const uid = new ObjectId().toHexString();
      const preset = (
        await PresetDal.addPreset(uid, {
          name: "first",
          settingGroups: ["hideElements"],
          config: {
            showKeyTips: true,
          },
        })
      ).presetId;

      //WHEN
      await PresetDal.editPreset(uid, {
        _id: preset,
        name: "newName",
        config: { ads: "off" },
      });

      //THEN
      const read = await PresetDal.getPresets(uid);
      expect(read[0]).toEqual(
        expect.objectContaining({
          _id: new ObjectId(preset),
          uid: uid,
          name: "newName",
          config: { ads: "off" },
          settingGroups: null,
        })
      );
    });
```

[the remark this answers is not in the document]
Comment (amarnathsama) [reply backend/__tests__/dal/preset.spec.ts]: fixed

**Comment by Miodec** [comment]
Thanks

**Pull request merged**